libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/entitlements/FilesEntitlement.java

```diff
@@ -41,6 +41,33 @@ public enum BaseDir {
 
     public sealed interface FileData {
 
+        sealed interface RelativeFileData extends FileData {
+            BaseDir baseDir();
+
+            Stream<Path> resolveRelativePaths(PathLookup pathLookup);
+
+            @Override
+            default Stream<Path> resolvePaths(PathLookup pathLookup) {
+                Objects.requireNonNull(pathLookup);
+                var relativePaths = resolveRelativePaths(pathLookup);
+                switch (baseDir()) {
+                    case CONFIG:
+                        return relativePaths.map(relativePath -> pathLookup.configDir().resolve(relativePath));
+                    case DATA:
+                        // multiple data dirs are a pain...we need the combination of relative paths and data dirs
+                        List<Path> paths = new ArrayList<>();
+                        for (var relativePath : relativePaths.toList()) {
+                            for (var dataDir : pathLookup.dataDirs()) {
+                                paths.add(dataDir.resolve(relativePath));
+                            }
+                        }
+                        return paths.stream();
+                    default:
+                        throw new IllegalArgumentException();
+                }
+            }
+        }
+
         final class AbsolutePathFileData implements FileData {
```

Comment:
Note: I'll align these implementations completely once https://github.com/elastic/elasticsearch/pull/122658 is merged.